### PR TITLE
Move definition of BIO_FLAGS_UPLINK completely to bio.h

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Move definition of BIO_FLAGS_UPLINK to bio.h

### CODE REVIEW

**Comment by bernd-edlinger** [comment]
Hmm...
You want to keep the definition of this macro dependent on which header file is included in which
sequence, for the next years to come?

**Comment by mspncp** [comment]
I think it should have a fixed value, not a variable one.

Grepping the source, I see the following in bio_lcl.h:

```
#if BIO_FLAGS_UPLINK==0
/* Shortcut UPLINK calls on most platforms... */
...
```

So it looks like the only obstacle for making that constant really constant is that it is (mis-)used for short-circuiting the applink calls. I recall that @dot-asm made some comment recently about short-circuiting in certain circumstances and that it doesn't work correctly anymore. But I don't recall the details and didn't find the location.  So maybe we should consult @dot-asm, he seems to be the only one who fully understands the applink beast...

**Comment by mspncp** [comment]
> I think it should have a fixed value, not a variable one.

And I agree with @levitte that it should not appear in public headers.

**Comment by mspncp** [comment]
I found it:  https://github.com/openssl/openssl/pull/7054#issuecomment-416046320

**Comment by bernd-edlinger** [comment]
If that helps, the public header could define BIO_FLAGS_UPLINK = 0,
which is used nowhere.
And internally we could rename BIO_FLAGS_UPLINK to BIO_FLAGS_UPLINK_INTERNAL.
And in we could remove BIO_FLAGS_UPLINK in the next major version,
whenever that will be.
Is there a way to deprecate a macro?

File: crypto/bio/bss_file.c [hunk 1]
```diff
@@ -142,13 +142,13 @@ static int file_read(BIO *b, char *out, int outl)
...
     if (b->init && (out != NULL)) {
-        if (b->flags & BIO_FLAGS_UPLINK)
+        if (b->flags & BIO_FLAGS_UPLINK_INTERNAL)
             ret = UP_fread(out, 1, (int)outl, b->ptr);
         else
             ret = fread(out, 1, (int)outl, (FILE *)b->ptr);
         if (ret == 0
-            && (b->flags & BIO_FLAGS_UPLINK) ? UP_ferror((FILE *)b->ptr) :
-                                               ferror((FILE *)b->ptr)) {
+            && (b->flags & BIO_FLAGS_UPLINK_INTERNAL)
```
Comment (bernd-edlinger) [inline crypto/bio/bss_file.c]: Funny, I think the brackets are misplaced here.
isn't this effectively `if ( (ret == 0 && b->flags & BIO_FLAGS_UPLINK) ? UP_ferror (b->ptr) : ferror(b->ptr) ) {`

**Comment by mspncp** [comment]
> Is there a way to deprecate a macro?

That's what Pauli told me to do in a similar situation:

https://github.com/openssl/openssl/commit/c402e943cd0d748ca2a74a37caeccdfc59ce2870#diff-57bbffc095bd596a230ff36b8d197d4aR27

**Comment by levitte** [comment]
I see no reason to rename it internally, just deprecate it publically.

**Comment by levitte** [comment]
Ah, so it's to ensure that none of our code uses the `bio.h` value...  Ok gotcha.

File: include/openssl/bio.h [hunk 2]
```diff
@@ -156,12 +156,9 @@ extern "C" {
 # define BIO_FLAGS_IO_SPECIAL    0x04
 # define BIO_FLAGS_RWS (BIO_FLAGS_READ|BIO_FLAGS_WRITE|BIO_FLAGS_IO_SPECIAL)
 # define BIO_FLAGS_SHOULD_RETRY  0x08
-# ifndef BIO_FLAGS_UPLINK
-/*
- * "UPLINK" flag denotes file descriptors provided by application. It
- * defaults to 0, as most platforms don't require UPLINK interface.
- */
-#  define BIO_FLAGS_UPLINK        0
+# if OPENSSL_API_COMPAT < 0x10200000L
```
Comment (mattcaswell) [inline include/openssl/bio.h]: This PR is marked for backport to 1.1.1, but we should not be making deprecation changes in the public headers in a stable release.

Comment (t8m) [reply include/openssl/bio.h]: True. So we just except the last commit from the cherry-pick?

Comment (mattcaswell) [reply include/openssl/bio.h]: Does this actually fix any bugs in 1.1.1? Or is it purely for cleanup reasons? If the latter then I'd say don't do it in 1.1.1 at all.

Comment (bernd-edlinger) [reply include/openssl/bio.h]: Mostly cleanup. yes.
However it is fragile to depend on including header files in a certain sequence.
It is even more fragile if master does not have this dependency while 1.1.1 has it.
BTW: Is API_COMPAT < 0x10200000L still correct, at that time I thought the next version would be 1.2.0,
but as we know now it is 3.0.0

Comment (mattcaswell) [reply include/openssl/bio.h]: Oh, good point. Should be:

````
# if !OPENSSL_API_3
````
I'm still minded to make this master only.

Comment (bernd-edlinger) [reply include/openssl/bio.h]: okay, then master only.

File: crypto/bio/bss_file.c [hunk 3]
```diff
@@ -142,13 +142,13 @@ static int file_read(BIO *b, char *out, int outl)
...
-        if (b->flags & BIO_FLAGS_UPLINK)
+        if (b->flags & BIO_FLAGS_UPLINK_INTERNAL)
             ret = UP_fread(out, 1, (int)outl, b->ptr);
         else
             ret = fread(out, 1, (int)outl, (FILE *)b->ptr);
         if (ret == 0
-            && (b->flags & BIO_FLAGS_UPLINK) ? UP_ferror((FILE *)b->ptr) :
-                                               ferror((FILE *)b->ptr)) {
+            && (b->flags & BIO_FLAGS_UPLINK_INTERNAL
+                ? UP_ferror((FILE *)b->ptr) : ferror((FILE *)b->ptr))) {
```
Comment (bernd-edlinger) [inline crypto/bio/bss_file.c]: Hmm, actually, *this* is a bug-fix.

**Comment by t8m** [comment]
ping?

**Comment by bernd-edlinger** [comment]
Merged to master as b113279.  Thanks!

**Pull request closed without merging**